hal/src/electron/modem/mdm_hal.cpp

```diff
@@ -1234,12 +1240,20 @@ bool MDMParser::getDataUsage(MDM_DataUsage &data)
 
 bool MDMParser::getCellularGlobalIdentity(CellularGlobalIdentity& cgi_) {
     bool result;
+    // Reformat the operator string to be numeric (allows the capture of `mcc` and `mnc`)
+    if (0 == sendFormated("AT+COPS=3,2\r\n")) {
```

Comment:
I still strongly discourage such a pattern in C/C++. Let's discuss this in a wider team and come to a unified stance on this if you feel that I'm simply being annoying, but it decreases the readability of code by a lot and also prevents the usage of RAII in-between the actions. cc @sergeuz @technobly @m-mcgowan

Reply:
I would love to have this conversation. To be honest, I don't fully understand your concern, and it would be nice to discuss it as opposed to messaging back and forth. I'm always happy to learn something new.

Reply:
I agree that an `if-else tree` should not be used as a state machine.  The `goto failure` pattern used in the rest of mdm_hal.cpp is more readable, or you can use a `do { } while (false);` block with `breaks` if you really despise goto's.

Reply:
The simplest example:
```cpp
int func() {
    CHECK(doSomething());
    CHECK(doSomethingElse());

    // Only at this point after doing something and something else
    // we want to for example acquire some lock
    RaiiLock lock;

    CHECK(doThirdThing());
    CHECK(doFourthThing());

    return 0;
}
```

How would you do this with your (if-else-waterfall? I don't know how to call it)?

I'm not accentuating the usage of `CHECK` macros here in the example, it can very well be a set of `ifs` with a `return` or a `goto` or the whole thing can be a `do-while` with `breaks`.

And it's not only about locking. With your pattern for example you are almost always required to declare all the variables at the top of the function, you cannot easily do so consecutively as the code flow goes which I find rather inconvenient and an atavism from C89.

Reply:
As for `goto`, Bjarne Stroustrup says it best in his talk, [What can C++ do for embedded systems developers?](https://youtu.be/VoHOLDdfDhk?t=3455)

> Why don't we use `goto`? The answer is, it can do _ANYTHING_. We want something better behaved...

Reply:
You are derailing conversation somewhat as both Brett and I have mentioned other possible ways to do this, but if you want to use Stroustrup's name as an argument point, please consider reading through [E: Error handling section of C++ Core Guidelines](https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#e-error-handling) also written by Stroustrup and specifically sections:

```
E.25: If you can’t throw exceptions, simulate RAII for resource management
E.26: If you can’t throw exceptions, consider failing fast
E.27: If you can’t throw exceptions, use error codes systematically
E.28: Avoid error handling based on global state (e.g. errno)
```

Reply:
I would wager, it is not as difficult to decipher as it is foreign (and therefore not immediately obvious).

1. It has the advantage of not being opaque (unlike the `CHECK` macro)
2. It can handle any result type (not only error values `< 0` or success codes `>= 0`
3. It can clean an non-RAII allocation from a previous step (allocation, lock, etc...)
   - Even if we perfect our codebase, our third-party dependencies can create this need.

Also, the example above may be symptomatic of an overly complex function, but if not, it can still be expressed. It requires one level of nesting at each RAII acquisition.

```cpp
int func() {
    int result;
    if (result = doSomething()) {
        // handle error condition
    } else if (result = doSomethingElse()) {
        // handle error condition and clean resources acquired during previous step(s)
    } else {
        // Only at this point after doing something and something else
        // we want to for example acquire some lock
        RaiiLock lock;

        if (result = doThirdThing()) {
            // handle error condition and clean resources acquired during previous step(s)
        } else if (result = doFourthThing()) {
            // handle error condition and clean resources acquired during previous step(s)
        } else {
            // do stuff
        }
    }
    return result;
}
```

Reply:
They can't do _ANYTHING_ in modern compilers.  Try to make a `goto` jump out of a function and into another function... see if that compiles ^_^  

>_Re: goto, The main reason you shouldn’t use them is because they make some people freak out. -Gramshackle_

Reply:
That actually looks at least somewhat palatable, due to simplicity of the example and lack of comments _between_ the blocks compared to the PR, but still, instead of going downwards, you need to go south-east in order to go through the flow of the function. Why? We can argue about short functions but even with such a simple example it's still apparent.

But again, this is example is trivial, you if add non-trivial C++ objects and other things into the equation, you'll end up nesting most of the times.